processing/src/main/java/io/druid/segment/ReferenceCountingSegment.java

```diff
@@ -19,23 +19,42 @@
 
 package io.druid.segment;
 
+import com.google.common.base.Preconditions;
 import com.metamx.emitter.EmittingLogger;
 import org.joda.time.Interval;
 
 import java.io.Closeable;
-import java.io.IOException;
+import java.util.concurrent.Phaser;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 public class ReferenceCountingSegment extends AbstractSegment
 {
   private static final EmittingLogger log = new EmittingLogger(ReferenceCountingSegment.class);
 
   private final Segment baseSegment;
-
-  private final Object lock = new Object();
-
-  private volatile int numReferences = 0;
-  private volatile boolean isClosed = false;
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+  private final Phaser referents = new Phaser(1)
+  {
+    @Override
+    protected boolean onAdvance(int phase, int registeredParties)
+    {
+      Preconditions.checkState(registeredParties == 0);
```

Comment:
Maybe this also should be in try block to suppress throwing an exception?

Reply:
this can only happen due to a bug in the code here, in that case its probably better to fail rather than succeed.